### PR TITLE
Implement Client::LockBucketRetentionPolicy.

### LINKED ISSUE #1029: Support Bucket Lock features.
The purpose of this issue is to track the initial work of finding out what those features are and creating smaller bugs to list the specific tasks to add support for these features.

### PR DESCRIPTION
**Pull request by coryan**
This function implements the `Buckets: lockRetentionPolicy` API. As
usual this PR also includes a unit test for the member function, an
integration test, and the code samples.  This fixes #1029.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1383)

### CODE REVIEW

File: google/cloud/storage/client.h [hunk 1]
```diff
@@ -459,6 +459,57 @@ class Client {
...
   }
+
+  /**
+   * Locks the retention policy for a bucket.
+   *
+   * @warning Locking a retention policy is an irreversible action. Once locked,
+   *     you must delete the entire bucket in order to "remove" the bucket's
+   *     retention policy. However, before you can delete the bucket, you must
+   *     be able to delete all the objects in the bucket, which itself is only
+   *     possible if the all objects have reached the retention period set by
```
Comment (frankyn) [inline google/cloud/storage/client.h]: `if all the object`

File: google/cloud/storage/client.h [hunk 2]
```diff
@@ -459,6 +459,57 @@ class Client {
...
+   *     the retention policy.
+   *
+   * The [Bucket Lock
+   * feature](https://cloud.google.com/storage/docs/bucket-lock) allows you to
+   * configure a data retention policy for a Cloud Storage bucket that governs
+   * how long objects in the bucket must be retained. The feature also allows
+   * you to lock the data retention policy, permanently preventing the policy
+   * from from being reduced or removed.
```
Comment (frankyn) [inline google/cloud/storage/client.h]: `from from` being reduced or removed only increased.

File: google/cloud/storage/client.h [hunk 3]
```diff
@@ -459,6 +459,57 @@ class Client {
...
+   *
+   * @throw std::runtime_error if the operation fails.
+   *
+   * @par Examples
+   *
+   * @snippet storage_bucket_samples.cc lock retention policy
+   *
+   * @snippet storage_bucket_samples.cc set retention policy
+   *
+   * @snippet storage_bucket_samples.cc remove retention policy
```
Comment (frankyn) [inline google/cloud/storage/client.h]: `@snippet storage_bucket_samples.cc get retention policy`

File: google/cloud/storage/idempotency_policy.cc [hunk 4]
```diff
@@ -263,6 +267,12 @@ bool StrictIdempotencyPolicy::IsIdempotent(
   return true;
 }
 
+bool StrictIdempotencyPolicy::IsIdempotent(
```
Comment (frankyn) [inline google/cloud/storage/idempotency_policy.cc]: Nice!

File: google/cloud/storage/client.h (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed.

File: google/cloud/storage/client.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed.

File: google/cloud/storage/client.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed.

File: google/cloud/storage/client.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/client.h]: Fixed.

File: google/cloud/storage/client.h [hunk 5]
```diff
@@ -459,6 +459,57 @@ class Client {
...
+   * @param options a list of optional query parameters and/or request headers.
+   *     Valid types for this operation include `UserProject`.
+   *
+   * @throw std::runtime_error if the operation fails.
+   *
+   * @par Examples
+   *
+   * @snippet storage_bucket_samples.cc lock retention policy
+   *
+   * @snippet storage_bucket_samples.cc set retention policy
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/client.h]: I am not sure.  @frankyn what would be the most reasonable order for this?  An alternative would be to move the set/get/remove retention policy examples to the `BucketMetada` class.

**Comment by coryan** [comment]
Addressed comments (thanks!) and rebased to resolve the conflict.
PTAL.

File: google/cloud/storage/client.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (houglum) [reply google/cloud/storage/client.h]: It seems more intuitive to me that those examples would be documented under their respective methods in the BucketMetadata class (maybe with a link/reference to those methods in this helptext?), but I'll defer to @frankyn.

[the remark this answers is not in the document]
Comment (frankyn) [reply google/cloud/storage/client.h]: It helps connect information revolving a feature. At least that's how I looked at it. Maybe I found lockRetentionPolicy but would like to know how to "get" or "set" the retention policy instead. Here's a link.

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/client.h]: Sounds good, you approve as-is then?

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/client.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (frankyn) [reply google/cloud/storage/client.h]: Yes.

**Pull request merged**